zilencer/management/commands/process_billing_updates.py

```diff
@@ -0,0 +1,73 @@
+"""\
+Run RealmAuditLogProcessors.
+
+This management command is run via supervisor. Do not run on multiple
+machines, as the code has not been made robust to race conditions from doing
+so.  (Alternatively, you can set `BILLING_UPDATER_DISABLED=True` on all but
+one machine to make the command have no effect.)
+
+"""
+
+import logging
+import time
+from typing import Any
+
+from django.conf import settings
+from django.core.management.base import BaseCommand
+from django.utils.timezone import now as timezone_now
+
+from zerver.lib.context_managers import lockfile
+from zerver.lib.logging_util import log_to_file
+
+from zilencer.lib.stripe import BILLING_LOG_PATH
+billing_logger = logging.getLogger('zilencer.stripe')
+log_to_file(billing_logger, BILLING_LOG_PATH)
+log_to_file(logging.getLogger('stripe'), BILLING_LOG_PATH)
+
+class Command(BaseCommand):
+    help = """Run RealmAuditLogProcessors, to sync billing-relevant updates into Stripe.
+
+Run this command under supervisor.
+
+Usage: ./manage.py process_billing_updates
+"""
+
+    def handle(self, *args: Any, **options: Any) -> None:
+        # TODO: make this a real setting
+        if settings.BILLING_UPDATER_DISABLED:
+            while True:
+                time.sleep(10*9)
+
+        with lockfile("/tmp/zulip_billing_updater.lockfile"):
+            while True:
+                updated = False
+                for processor in RealmAuditLogProcessor.objects \
+                                                       .exclude(state=RealmAuditLogProcessor.STALLED):
+                    try:
+                        if process_next_billing_log_entry(processor):
+                            updated = True
+                        else:
+                            pass  # TODO: delete processor if processor.realm is not None
+                    except error:
+                        if is_customer_agnostic_error(error):
+                            # log error
+                            time.sleep(2*60)
+                        else:
+                            if processor.realm is None:
+                                # log error
+                                RealmAuditLogProcessor.objects.create(
+                                    log_row=processor.log_row,
+                                    realm=processor.log_row.realm,
+                                    state=RealmAuditLogProcessor.STALLED)
+                                processor.state = RealmAuditLogProcessor.DONE
+                                processor.save()
+                            else:
+                                # log error
+                                processor.state = RealmAuditLogProcessor.STALLED
+                                processor.save()
```

Comment:
Why are there two cases here?  I guess it's not clear to me why `processor.realm` would not exist, but `processor.log_row.realm` would.

Reply:
`processor.realm is None` is the global processor. all `processor.log_row`s are guaranteed to have a `realm`

Reply:
OK.  I think I was just confused by some of the wrong-model bugs elsewhere making the code hard to read, but I also think this could be made a lot more naturally readable by copying some of the content you have in the design doc into comments in this file.